proposals/kapp/001-crd-upgrade-safety/README.md

```diff
@@ -0,0 +1,95 @@
+## Problem Statement
+CustomResourceDefinitions (CRDs) are an integral part of extending the Kubernetes API and are often applied to a cluster alongside a controller that manages them.
+
+Since CRDs are global and only one instance can exist on the cluster at a time, it is important to ensure that updates to a CRD are performed safely.
+This means preventing data loss and breaking of clients/workloads that may be using the API.
+Failure to safely update a CRD may result in:
+- Existing CustomResources (CRs) being invalidated, causing extensions to fail reconciliation of the existing CRs
+- Breaking clients expecting the old schema to be present. This may include other extensions, workloads, and tooling that depend on a specific version of the CRD.
+
+When an unsafe update occurs, it requires manual intervention to update all of the existing CRs to the new version of the CRD.
+
+## Terminology / Concepts
+- CustomResourceDefinition (CRD) - https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/
+- CustomResource (CR) - https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/
+- Fail Closed - In the scope of this document, "failing closed" is used to represent that if we can't explicitly validate something then we should fail.
+- Fail Open - In the scope of this document, "failing open" is used to represent that if we can't explicitly validate something then we should succeed.
+- Preflight Check - A preflight check is a validation that is run prior to performing some action. Typically, if the validation fails the action will not be performed.
+
+## Proposal
+In order to address the problem identified, it is proposed that a new, optional, preflight check is added to `kapp` to enable verifying that an `update` operation on a CRD is safe before attempting to apply any changes to the cluster.
+
+### Goals and Non-goals
+
+#### Goals
+- Providing `kapp` users a way to enable verification that an `update` operation on CRDs is safe and won't cause data loss or break clients
+
+#### Non-goals
+- Anything not pertaining to providing `kapp` users a way to ensure that `update` operations on CRDs are safe before applying changes
+
+### Specification
+A new preflight check will be added to `kapp` called `CRDUpgradeSafety`.
+
+#### `preflight` Package Changes
+The `CRDUpgradeSafety` preflight check is a very nuanced preflight check and different users will likely want to be able to configure this check a bit further to do things like:
+- Only issue warnings about potential problems
+- Fail closed/open
+
+In order to accomodate this configurability without introducing various versions of the same preflight check, the logic in https://github.com/carvel-dev/kapp/tree/develop/pkg/kapp/preflight will be updated to allow for configurable preflight checks.
+
+As an example, configuring a preflight check could look something like:
+```sh
+kapp deploy ... --preflight=Check=true:{"mode":"strict"}
+```
```

Comment:
This format causes problems if we expect to have more than one check, and multiple JSON values:
```
--preflight=Check=true:{"mode":"strict","bob":"alice"},Check2=true:{"mode":"open","bob":"carol"}
```
Too many commas and colons to parse easily. I'm thinking either it should be all JSON, or different delimiters should be used...
e.g.
```
--preflight=Check1={"enabled":"true"};Check2={"enabled":"true","mode":"strict"}
```
or
```
--preflight={"Check1":{"enabled":"true"},"Check2":{"enabled":"true","mode":"strict"}}
```

Reply:
Updated the configuration section to instead recommend using the existing kapp configuration file method in https://github.com/carvel-dev/carvel/pull/729/commits/5d56b1f645a139ac817062bb026c4ea77617e146

Reply:
I was about to point out that if it's JSON in the command-line, then the examples should single-quote it, otherwise your common shell will strip the quotes. And it's so finicky to get JSON right inside a command-line arg.

```
❯ cat --style=plain ./test.sh
#!/usr/bin/env bash

echo "$@"

❯ ./test.sh --preflight=Check=true:{"mode":"strict"}
--preflight=Check=true:{mode:strict}

❯ ./test.sh --preflight=Check=true:'{"mode":"strict"}'
--preflight=Check=true:{"mode":"strict"}

```

However,➕ to @joaopapereira's point on moving it to `kind: Config`.